lua/ttt2/libraries/modelbuilder.lua

```diff
@@ -0,0 +1,316 @@
+-- @realm client
+function modelbuilder.CreateModel(wep, modelData)
+	-- handle a model being added to the view or world model
+	if modelData.type == "Model" and modelData.model and modelData.model ~= ""
+		and (not IsValid(modelData.modelEnt) or modelData.createdModel ~= modelData.model)
+		and string.find(modelData.model, ".mdl") and file.Exists (modelData.model, "GAME")
+	then
+		modelData.modelEnt = ClientsideModel(modelData.model, RENDER_GROUP_VIEW_MODEL_OPAQUE) --todo check correct render group
+
+		if IsValid(modelData.modelEnt) then
+			modelData.modelEnt:SetPos(wep:GetPos())
+			modelData.modelEnt:SetAngles(wep:GetAngles())
+			modelData.modelEnt:SetParent(wep)
+			modelData.modelEnt:SetNoDraw(true)
+			modelData.createdModel = modelData.model
+		else
+			modelData.modelEnt = nil
+		end
+
+	-- handle a sprite being added to the view or world model
+	elseif modelData.type == "Sprite" and modelData.sprite and modelData.sprite ~= ""
+		and (not modelData.spriteMaterial or modelData.createdSprite ~= modelData.sprite)
+		and file.Exists ("materials/" .. modelData.sprite .. ".vmt", "GAME")
+	then
+		local name = modelData.sprite .. "-"
+		local params = {
+			["$basetexture"] = modelData.sprite
+		}
+
+		-- make sure we create a unique name based on the selected options
+		for i = 1, #propertiesMaterial do
+			local property = propertiesMaterial[i]
+
+			if modelData[property] then
+				params["$" .. property] = 1
+				name = name .. "1"
+			else
+				name = name .. "0"
+			end
+		end
+
+		modelData.createdSprite = modelData.sprite
+		modelData.spriteMaterial = CreateMaterial(name, "UnlitGeneric", params)
+	end
+end
+
+---
+-- Resets the bone positions of a view model.
+-- @param Entity viewModel The view model of the weapon
+-- @realm client
+function modelbuilder.ResetBonePositions(viewModel)
+	local boneCount = viewModel:GetBoneCount()
+
+	if not boneCount then return end
+
+	for i = 0, boneCount do
+		viewModel:ManipulateBoneScale(i, Vector(1, 1, 1))
+		viewModel:ManipulateBoneAngles(i, Angle(0, 0, 0))
+		viewModel:ManipulateBonePosition(i, Vector(0, 0, 0))
+	end
+end
+
+---
+-- Updates the bone positions of the view model of a weapon. Is used to hide a
+-- default view model of a weapon.
+-- @param Entity wep The weapon for which the view model should be updated
+-- @param Entity viewModel The view model of the weapon
+-- @realm client
+function modelbuilder.UpdateBonePositions(wep, viewModel)
+	if not wep.ViewModelBoneMods then
+		modelbuilder.ResetBonePositions(viewModel)
+
+		return
+	end
+
+	local allBones = {}
+
+	for i = 0, viewModel:GetBoneCount() do
+		local bonename = viewModel:GetBoneName(i)
+
+		if wep.ViewModelBoneMods[bonename] then
+			allBones[bonename] = wep.ViewModelBoneMods[bonename]
+		else
+			allBones[bonename] = {
+				scale = Vector(1, 1, 1),
+				pos = Vector(0, 0, 0),
+				angle = Angle(0, 0, 0)
+			}
+		end
+	end
+
+	for identifier, dataTable in pairs(allBones) do
```

Comment:
this loop could be combined with the loop above i think

Reply:
I did this, I think it works, but I could not test it